File: typescript/src/connection.ts
/**
 * RpcConnection — composes a MessageRouter with an ITransportConnection.
 * Used on the server side (one per connected peer) and as a standalone
 * bidirectional RPC channel.
 */

import type { EventCallback, RpcHandler } from "./core.js";
import { MessageRouter } from "./router.js";
import type { ITransportConnection } from "./transport.js";

export class RpcConnection {
	readonly router: MessageRouter;
	readonly transport: ITransportConnection;

	/** Arbitrary metadata (role, client_id, token, authenticated, etc.) */
	meta: Record<string, unknown> = {};

	constructor(transport: ITransportConnection, opts?: { timeout?: number }) {
		this.transport = transport;
		this.router = new MessageRouter(
			(raw) => this.transport.send(raw),
			opts?.timeout ?? 30_000,
		);

		// Wire pong callback to transport
		this.router.onPong = () => {
			if ("refreshPong" in this.transport) {
				(this.transport as { refreshPong(): void }).refreshPong();
			}
		};
	}

	get isOpen(): boolean {
		return this.transport.isOpen;
	}

	// ── Lifecycle ────────────────────────────────────────────────────────

	/**
	 * Start the message loop. Resolves when the connection closes.
	 * Used on the server side to await connection lifetime.
	 */
	serve(): Promise<void> {
		return new Promise<void>((resolve) => {
			this.transport.onMessage = (raw) => this.router.dispatchMessage(raw);
			this.transport.onClose = () => {
				this.router.close();
				resolve();
			};

			// If the transport has a serve() method (e.g. WsConnection), call it
			// eslint-disable-next-line @typescript-eslint/no-explicit-any
			const t = this.transport as any;
			if (typeof t.serve === "function") {
				t.serve().then(() => {
					this.router.close();
					resolve();
				});
			}
		});
	}

	close(): void {
		this.router.close();
		this.transport.close();
	}

	// ── RPC methods (delegate to router) ────────────────────────────────

	register(method: string, handler: RpcHandler): void {
		this.router.register(method, handler);
	}

	unregister(method: string): void {
		this.router.unregister(method);
	}

	request<T = unknown>(
		method: string,
		params?: unknown,
		timeoutMs?: number,
	): Promise<T> {
		return this.router.request<T>(method, params, timeoutMs);
	}

	batchRequest<T = unknown>(
		calls: Array<[method: string, params?: unknown]>,
		timeoutMs?: number,
	): Promise<T[]> {
		return this.router.batchRequest<T>(calls, timeoutMs);
	}

	// ── Pub/Sub (delegate to router) ────────────────────────────────────

	publish(method: string, params?: unknown): void {
		this.router.publish(method, params);
	}

	subscribe(method: string, callback: EventCallback): void {
		this.router.subscribe(method, callback);
	}

	unsubscribe(method: string, callback: EventCallback): void {
		this.router.unsubscribe(method, callback);
	}
}

File: python/src/echorpc/connection.py
"""RpcConnection — composes a MessageRouter with a transport connection."""

from __future__ import annotations

import asyncio
from typing import Any

from .core import DEFAULT_REQUEST_TIMEOUT
from .router import EventCallback, Handler, MessageRouter


class RpcConnection:
    """Wraps a transport connection and provides bidirectional JSON-RPC 2.0."""

    def __init__(
        self,
        transport: Any,
        *,
        timeout: float = DEFAULT_REQUEST_TIMEOUT,
    ) -> None:
        self.transport = transport
        self.meta: dict[str, Any] = {}
        self._close_event = asyncio.Event()

        self.router = MessageRouter(
            lambda raw: self.transport.send(raw),
            timeout=timeout,
        )

        # Wire pong callback to transport
        self.router.on_pong = self._on_pong

    @property
    def is_open(self) -> bool:
        return self.transport.is_open

    # ── Lifecycle ────────────────────────────────────────────────────────

    async def serve(self) -> None:
        """Start message loop. Blocks until connection closes."""
        self.transport.on_message = self._on_message
        self.transport.on_close = self._on_close

        # If transport has serve() (e.g. WsConnection), use it to block
        if hasattr(self.transport, "serve"):
            await self.transport.serve()
        else:
            # For transports without serve() (e.g. HTTP), wait for close event
            await self._close_event.wait()

    async def close(self) -> None:
        self.router.close()
        await self.transport.close()

    def _on_message(self, raw: str) -> Any:
        return self.router.dispatch_message(raw)

    def _on_close(self) -> None:
        self.router.close()
        self._close_event.set()

    def _on_pong(self) -> None:
        if hasattr(self.transport, "refresh_pong"):
            self.transport.refresh_pong()

    # ── RPC methods (delegate to router) ────────────────────────────────

    def register(self, method: str, handler: Handler) -> None:
        self.router.register(method, handler)

    def unregister(self, method: str) -> None:
        self.router.unregister(method)

    async def request(
        self, method: str, params: Any = None, *, timeout: float | None = None
    ) -> Any:
        return await self.router.request(method, params, timeout=timeout)

    async def batch_request(
        self,
        calls: list[tuple[str, Any]],
        *,
        timeout: float | None = None,
    ) -> list[Any]:
        return await self.router.batch_request(calls, timeout=timeout)

    # ── Pub/Sub (delegate to router) ────────────────────────────────────

    async def publish(self, method: str, params: Any = None) -> None:
        await self.router.publish(method, params)

    def subscribe(self, method: str, callback: EventCallback) -> None:
        self.router.subscribe(method, callback)

    def unsubscribe(self, method: str, callback: EventCallback) -> None:
        self.router.unsubscribe(method, callback)
